### DISCUSSION TITLE
println vs trailing comma?

### DISCUSSION THREAD

**Discussion by waywardgeek** [google/rune#40] [Polls]
Should we drop the println statement, and instead always print a newline, unless a print statement ends with a trailing comma?  This would be more like Python 2.X, and less like Pascal.

Folks new to a language like to see what happens when they write:

```python
print "Hello, World!"
```

It is nice when this just works, like it did in Python 2.X.  It is also nicer to use when debugging.  This works in many languages including BASIC.  What if we don't want to print a newline?  In Python 2.X, similar to BASIC where they use a trailing semicolon, a trailing comma means don't write the newline.

One of the first languages I learned was Pascal.  That's why we have print and println as separate statements in Rune.  It is also why we use the 'var' keyword to pass a pointer to a parameter to a function, rather than the parameter's value.  However, Rune is supposed to be more like Python, rather than Pascal.  I feel like the var keyword is justified because Rune is a systems programming language where passing a pointer to a value rather than the value can make code faster, which is a non-goal in Python.  However, println vs a trailing comma seems like it has a weak argument for being different than Python.

Should we switch?

Bill

**Comment by 0dminnimda** [comment]
I am not sure if I fully understood the problem, but why not use `"\n"` and just have `print`?

**Comment by benji-york** [comment]
I remember Tim Peters one time saying that syntax shouldn't look like grit on his screen.  (I.e., visually small changes to code should not result in a significant change in semantics.)

I think that applies here, the visual difference between `print "Hello, World!"` and `print "Hello, World!",` is very small.

**Reply by waywardgeek** [reply]
For a low-vision developer, your argument carries a lot of weight.  OK, so no trailing commas...  Then what?  I am tempted to go with the suggestion above, adding "\n" when needed, which matches C/C++, and some others.  However, it doesn't quite match Python.

One gripe I've had with Python's print statement in both 2.X and 3.X is that it can't easily be used by new programmers for formatting text output.  There is always a space between elements, which makes printing formatted text hard.  Rune already differes from Python in not printing those spaces.  Maybe diverge even further and just use the \n manually as suggested above?

**Comment by 0dminnimda** [comment]
What is the language priority, being like python? Then make a library function
If not, then I see the current print with no spaces in between the parameters as a good enough alternative

`print("a, b, c:", a, b, c)` == `print "a, b, c: ", a, " ", b, " ", c, "\n"`

Although there is an issue of excess quotes, and it feels like making string interpolation a default behavior is a positive change and a thing language can settle down with

`print("a, b, c:", a, b, c)` == `print "a, b, c: {a} {b} {c}\n"` <- `print "a, b, c: ", a, " ", b, " ", c, "\n"`

For starters this feature can be as primitive as only accepting variables, no complex expressions, it's slightly inconvenient, but really improves the readability of the code

**Reply by waywardgeek** [reply]
Rune does not have variable parameters, yet.  This is why print statements exist.  I suspect this is also why print was a statement in the original Python.

**Reply by 0dminnimda** [reply]
nim solves this problem by having echo which always adds new line and a general write function for steams - `stdout.write x` vs `echo x`
Not sure if this is a good design, but `print` with new line and `write` without can make it look more like python

**Reply by 0dminnimda** [reply]
Implementing variable parameters (or arguments, for my taste) is not the easiest and I guess not the most important task right now, so you can just rename `printLn` -> `print`, `print` -> `write` and forget about it for some time

There are just not that many cases when you **need** to wirte without adding a new line, so this seems like an optimal solution
